### PR TITLE
feat(document): add flattenUUIDs option to `toObject()` and `toJSON()`

### LINKED ISSUE #15021: flattenUUIDs
Re: https://github.com/Automattic/mongoose/discussions/14969#discussioncomment-11005437

### PR DESCRIPTION
**Pull request by AbdelrahmanHafez**
## Summary
  - Add `flattenUUIDs` option to `toObject()` and `toJSON()` that converts UUID instances to 36-character hex strings, similar to `flattenObjectIds`
  - Add TypeScript support with `UUIDToString<T>` type helper and method overloads
  - Rename `UUIDToJSON` to `UUIDToString` for consistency with `ObjectIdToString`, keeping `UUIDToJSON` as a deprecated alias for backwards compatibility

Fixes #15021

### CODE REVIEW

**Review by hasezoey — APPROVED** [review]

File: types/document.d.ts [hunk 1]
```diff
@@ -287,6 +287,10 @@ declare module 'mongoose' {
     toJSON(options: ToObjectOptions & { flattenMaps: true, virtuals: true }): FlattenMaps<Require_id<DocType & TVirtuals>>;
     toJSON(options: ToObjectOptions & { flattenMaps: true }): FlattenMaps<Require_id<DocType>>;
 
+    // flattenUUIDs overloads
+    toJSON(options: ToObjectOptions & { flattenUUIDs: true, virtuals: true }): UUIDToString<Require_id<DocType & TVirtuals>>;
```
Comment (vkarpov15) [inline types/document.d.ts]: Unfortunately we would also need to handle all other combinations:

- `flattenUUIDs: true, flattenMaps: true`
- `flattenUUIDs: true, flattenObjectIds: true`
- `flattenUUIDs: true, flattenObjectIds: true, flattenMaps: true`
- `flattenUUIDs: true, flattenObjectIds: true, flattenMaps: true, virtuals: true`
- `flattenUUIDs: true, flattenObjectIds: true, virtuals: true`
- `flattenUUIDs: true, flattenMaps: true, virtuals: true`

Plus versionKey variants for completeness' sake.

Honestly this is why I don't like adding new options to toObject() and toJSON() unless absolutely necessary: they cause combinatorial explosion in TypeScript. There isn't a good way to handle all the different cases without explicitly listing out every single override.

File: types/index.d.ts [hunk 2]
```diff
@@ -927,23 +929,33 @@ declare module 'mongoose' {
           } : T;
 
     /**
-    * Converts any Buffer properties into "{ type: 'buffer', data: [1, 2, 3] }" format for JSON serialization
+    * Converts any UUID properties into strings for JSON serialization
     */
-    export type UUIDToJSON<T> = T extends mongodb.UUID
+    export type UUIDToString<T> = T extends Types.UUID
```
Comment (vkarpov15) [inline types/index.d.ts]: Why is rename necessary?

**Review by vkarpov15 — CHANGES_REQUESTED** [review]
Need to make sure the TypeScript types are complete for handling all possible options overrides, which is a pain but it's the right thing to do. Adding all these overrides makes me wonder whether adding this feature is even worthwhile :worried:

File: types/index.d.ts (hunk 2 above)
Comment (AbdelrahmanHafez) [reply types/index.d.ts]: The rename isn't strictly necessary. I did it for consistency with `ObjectIdToString` (since we have `flattenObjectIds` -> `ObjectIdToString`, it felt natural to have `flattenUUIDs` -> `UUIDToString`). The old `UUIDToJSON` name is preserved as a deprecated alias for backwards compatibility.

That said, I'm happy to drop the rename entirely if you'd prefer to keep things simpler.

File: types/document.d.ts (hunk 1 above)
Comment (AbdelrahmanHafez) [reply types/document.d.ts]: I reworked this to use compositional conditional types instead of explicit overloads for every combination:

```typescript
type ApplyVirtuals<T, TVirtuals, O> = O extends { virtuals: true } ? T & TVirtuals : T;
type ApplyVersionKey<T, O, TSchemaOptions> = O extends { versionKey: false } ? Omit<T, VersionKeyName> : T;
type ApplyFlattenTransforms<T, O> = // handles flattenMaps, flattenObjectIds, flattenUUIDs in one pass
```

These feed into `ToObjectReturnType<DocType, TVirtuals, O, TSchemaOptions>` which composes them all together. Adding a new option now just means adding one more transform type instead of doubling the overload count, so future options won't blow up the type definitions.

The flatten options (`flattenMaps`, `flattenObjectIds`, `flattenUUIDs`) are handled by a single recursive type that processes all three in one traversal. I initially had separate recursive types for each, but nesting them caused TypeScript's "union type too complex" error. The single-pass approach avoids that and also correctly handles edge cases like ObjectIds nested inside Map values.

File: test/types/schema.test.ts [hunk 3]
```diff
@@ -2015,7 +2015,7 @@ function testInferHydratedDocTypeFromSchema() {
     map?: Map<string, string> | null | undefined
   }, { id: string }, {}, { id: string }>;
 
-  expectType<Expected>({} as HydratedDocType);
+  expectAssignable<Expected>({} as HydratedDocType);
```
Comment (AbdelrahmanHafez) [inline test/types/schema.test.ts]: Changed to `expectAssignable` due to how TypeScript compares type identity.

This test compares two Document types:
- `HydratedDocType`: inferred from the schema, has `TSchemaOptions = ResolveSchemaOptions<{}>`
- `Expected`: manually constructed, has `TSchemaOptions = DefaultSchemaOptions`

Even though this test never calls `toJSON`, `expectType` checks for type **identity** by comparing the entire structure of both types, including all method signatures.

Before this PR, `TSchemaOptions` wasn't used in any Document method signatures:
```typescript
toJSON(options?: ToObjectOptions): Require_id<DocType>;
```

After this PR, `TSchemaOptions` appears in return types:
```typescript
toJSON(options?: ToObjectOptions): Default__v<Require_id<DocType>, TSchemaOptions>;
```

So now TypeScript sees different method signatures on the two Document types:
```typescript
// HydratedDocType's toJSON:
toJSON(): Default__v<..., ResolveSchemaOptions<{}>>

// Expected's toJSON:
toJSON(): Default__v<..., DefaultSchemaOptions>
```

`ResolveSchemaOptions<{}>` and `DefaultSchemaOptions` are structurally identical (they resolve to the same shape), but they're different type expressions. This makes the entire Document types non-identical, even though they're fully compatible.

`expectAssignable` is correct here because the types ARE compatible for all practical purposes.

File: types/inferschematype.d.ts [hunk 4]
```diff
@@ -112,7 +112,12 @@ declare module 'mongoose' {
       }[alias]
     : unknown;
 
-  type ResolveSchemaOptions<T> = MergeType<DefaultSchemaOptions, T>;
+  type TypesAreEqual<A, B> = [A] extends [B] ? [B] extends [A] ? true : false : false;
+
+  type ResolveSchemaOptions<T> =
```
Comment (AbdelrahmanHafez) [inline types/inferschematype.d.ts:115]: Updated `ResolveSchemaOptions` to preserve type identity when `T` is `{}` or `DefaultSchemaOptions`.

When creating a schema without options (`new Schema({...})`), TypeScript infers `TSchemaOptions` as `{}`. Previously, `ResolveSchemaOptions<{}>` returned `MergeType<DefaultSchemaOptions, {}>`, which is structurally identical to `DefaultSchemaOptions` but a different type expression. This caused `expectType` checks to fail when comparing inferred document types against manually constructed ones.

Now `ResolveSchemaOptions` returns `DefaultSchemaOptions` directly in these cases, so the types are truly identical.

Without this change we need to modify existing tests, see https://github.com/Automattic/mongoose/pull/15864#discussion_r2640762977

File: test/types/document.test.ts [hunk 5]
```diff
@@ -652,6 +660,175 @@ async function gh15578() {
...
+function testObjectIdsInsideMaps() {
+  // Test that ObjectIds/UUIDs nested inside Map values are correctly converted
+  interface DocWithMapOfObjectIds {
+    _id: Types.ObjectId;
+    userRefs: Map<string, { oderId: Types.ObjectId }>;
+    uuidRefs: Map<string, { refId: Types.UUID }>;
+  }
+
+  const schema = new Schema<DocWithMapOfObjectIds>({
+    userRefs: { type: Map, of: { oderId: Schema.Types.ObjectId } },
```
Comment (vkarpov15) [inline test/types/document.test.ts]: Minor typo - this could be `orderId` or `otherId`?

File: types/index.d.ts (hunk 2 above)
Comment (vkarpov15) [reply types/index.d.ts]: I'd prefer to keep the old name, at least unless we're going to drop support for the old name in the next major release. I'd prefer to avoid adding aliased names unless there's a strong reason to - keep the API consistent

File: types/inferschematype.d.ts [hunk 6]
```diff
@@ -112,7 +112,12 @@ declare module 'mongoose' {
       }[alias]
     : unknown;
 
-  type ResolveSchemaOptions<T> = MergeType<DefaultSchemaOptions, T>;
+  type TypesAreEqual<A, B> = [A] extends [B] ? [B] extends [A] ? true : false : false;
```
Comment (vkarpov15) [inline types/inferschematype.d.ts]: This helper should be equivalent to `IfEquals<>`, can we reuse IfEquals to avoid adding a separate helper?

File: types/index.d.ts [hunk 7]
```diff
@@ -1011,14 +1023,85 @@ declare module 'mongoose' {
...
+    // Handle regular objects - recurse into properties
+    : T extends Record<string, any>
+      ? { [K in keyof T]: ApplyFlattenTransforms<T[K], O> }
+    : T;
+
+  /**
+   * Computes the return type of toObject/toJSON based on the provided options.
+   * Uses a single-pass transform for flatten operations to correctly handle all combinations.
+   */
+  export type ToObjectReturnType<DocType, TVirtuals, O extends ToObjectOptions, TSchemaOptions = {}> =
```
Comment (vkarpov15) [inline types/index.d.ts:1086]: image

Comment (AbdelrahmanHafez) [reply types/index.d.ts:1086]: 🫡

**Comment by vkarpov15** [comment]
Merging into 9.2.
